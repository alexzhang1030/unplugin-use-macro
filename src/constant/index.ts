export const VALID_ID_RE = /.([tj]sx?|vue|svelte)$/
export const DEFINE_MACRO = 'useMacro'
