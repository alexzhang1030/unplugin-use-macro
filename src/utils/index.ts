import type { NodePath } from '@babel/core'
import type { Identifier } from '@babel/types'
import { DEFINE_MACRO, VALID_ID_RE } from '../constant'

export const isValidId = (id: string) => {
  return VALID_ID_RE.test(id)
}

export const isDefineMacro = (path: NodePath<Identifier>) => {
  return path.node.name === DEFINE_MACRO
}

export const getBodyAndName = (funcArgus: NodePath[]) => {
  const [name, body] = funcArgus
  if (!name.isStringLiteral() || !body.isArrowFunctionExpression())
    throw new Error('Invalid defineMacro arguments')
  return {
    name, body,
  }
}
