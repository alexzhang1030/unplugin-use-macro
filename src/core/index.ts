import type { NodePath } from '@babel/core'
import { parseSync, traverse } from '@babel/core'
import type { ArrowFunctionExpression } from '@babel/types'
import Ms from 'magic-string'
import { getBodyAndName, isDefineMacro } from '../utils'

export const Macros = new Map<string, NodePath<ArrowFunctionExpression>>()

export function parse(code: string) {
  let insertPos = 0
  const ast = parseSync(code)
  const codeMs = new Ms(code)
  traverse(ast, {
    CallExpression(path) {
      const callee = path.get('callee')
      if (callee.isIdentifier() && isDefineMacro(callee)) {
        const { name, body } = getBodyAndName(path.get('arguments'))
        if (!Macros.has(name.node.value))
          Macros.set(name.node.value, body)
        else
          console.warn(`Duplicate macro define: ${name}`)
        codeMs.remove(path.node.start!, path.node.end!)
        insertPos = path.node.start!
      }
    },
    LabeledStatement(path) {
      const label = path.get('label')
      const name = label.node.name
      if (!name.endsWith('$'))
        return
      const macroName = name.slice(0, -1)
      const macroBody = Macros.get(macroName)
      codeMs.remove(path.node.start!, path.node.end!)
      const callMacroBody = path.get('body')
      if (macroBody) {
        const params = callMacroBody.toString().replaceAll('\n', '').slice(1, -2).trim()
        codeMs.appendRight(insertPos, `
        (${macroBody?.toString()})(${params})
      `)
      }
    },
  })
  return codeMs.toString()
}
