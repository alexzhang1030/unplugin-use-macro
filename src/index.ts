import { createUnplugin } from 'unplugin'
import { parse } from './core'
import type { Options } from './types'
import { isValidId } from './utils'

export default createUnplugin<Options | undefined>(() => ({
  name: 'unplugin-use-macro',
  transformInclude(id) {
    return isValidId(id)
  },
  transform(code) {
    return parse(code)
  },
}))
