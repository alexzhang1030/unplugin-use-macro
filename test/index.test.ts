import { describe, expect, it } from 'vitest'
import { parse } from '../src/core'
import { isValidId } from '../src/utils'

describe('valid id', () => {
  it('valid id', () => {
    const data = ['ts', 'tsx', 'js', 'jsx', 'vue', 'svelte', '.png']
    for (const item of data)
      expect(isValidId(`index.${item}`)).toMatchSnapshot(item)
  })
  it('test', () => {
    const code = `
      useMacro("log", ($1, $2) => {
        console.log($2, $1)
      })

      log$: {
        "hello world",
        "next world"
      };
    `
    expect(parse(code)!).toMatchInlineSnapshot(`
      "
            

            ;
          
              (($1, $2) => {
        console.log($2, $1);
      })(\\"hello world\\", \\"next world\\")
            "
    `)
  })
})
