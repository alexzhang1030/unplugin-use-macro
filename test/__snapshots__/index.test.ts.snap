// Vitest Snapshot v1

exports[`valid id > valid id > .png 1`] = `false`;

exports[`valid id > valid id > js 1`] = `true`;

exports[`valid id > valid id > jsx 1`] = `true`;

exports[`valid id > valid id > svelte 1`] = `true`;

exports[`valid id > valid id > ts 1`] = `true`;

exports[`valid id > valid id > tsx 1`] = `true`;

exports[`valid id > valid id > vue 1`] = `true`;
